### PR TITLE
Fixed missing required wallet id and account name

### PR DESCRIPTION
**Pull request by randomlogin**
Previously `mkwallet` could be invoked without wallet id, which will become 'null', the second invocation results in error.
Same for wallet account creation.

- errors thrown if these parameters are missing
- sorted alphabetically help hint
- added hint about passing wallet id
- added 'admin' hints (for example about wallet creation) into the 'default' ones

### CODE REVIEW

**Comment by randomlogin** [comment]
Partially connected with: #24

**Comment by pinheadmz** [comment]
Nice thanks! Reviewing now. See related: https://github.com/bcoin-org/bcoin/issues/718

**Review by pinheadmz — CHANGES_REQUESTED** [review]
Thanks for this -- So, I think all the `admin` stuff in this file is outdated and seems like it can be basically removed. Look how we force-add the `admin` modifier just to pop it off again:

https://github.com/handshake-org/hs-client/blob/d82611cead27d7f9586d878438a23ad2005ff8af/bin/hsw-cli#L637-L651

I think we should remove all the admin crap and just merge the help list of commands all into one, like you have, but also merge the commands in the two `switch` statements. 

To assist the users, the commands that require `admin-token` if `wallet-auth=true` should be marked somehow, either by a `*` or by being separated to a sub-list at the end of the help command, perhaps with a link to https://hsd-dev.org/api-docs/#wallet-admin-commands

It would also be great to go one step further with this, port https://github.com/bcoin-org/bcoin/pull/705 from bcoin over to hsd, and then include a message in CLI that was like `"Execute 'hsw-rpc help' for more commands"` or something.... thoughts on that?

File: bin/hsw-cli [hunk 1]
```diff
@@ -595,39 +595,44 @@ class CLI {
...
+        this.log('  $ listen: Listen for events.');
+        this.log('  $ lock: Lock wallet.');
+        this.log('  $ master: View wallet master key.');
+        this.log('  $ mktx [address] [value]: Create transaction.');
+        this.log('  $ mkwallet [id]: Create wallet.');
+        this.log('  $ pending: View pending TXs.');
+        this.log('  $ rescan [height]: Rescan for transactions.');
+        this.log('  $ resend: Resend pending transactions.');
         this.log('  $ retoken: Create new api key.');
+        this.log('  $ rpc [command] [args]: Execute RPC command.')
```
[the remark this answers is not in the document]
Comment (randomlogin) [reply bin/hsw-cli:623]: added

File: lib/wallet.js [hunk 2]
```diff
@@ -197,6 +197,9 @@ class WalletClient extends Client {
    */
 
   createWallet(id, options) {
+    if (id == null) {
```
[the remark this answers is not in the document]
Comment (randomlogin) [reply lib/wallet.js:200]: added

**Comment by randomlogin** [comment]
> "Execute 'hsw-rpc help' for more commands" or something.... thoughts on that?
I don't get it, now if you call `hsw-cli rpc help sendbid` it shows the help.
If you don't provide a command it unrecognizes a command and shows the whole hint.

File: bin/hsw-cli [hunk 3]
```diff
@@ -592,93 +592,76 @@ class CLI {
...
         break;
+      case 'wallets':
+        await this.getWallets();
+        break;
+      case 'mkwallet':
+        await this.createWallet();
+        break;
+      case 'resend':
+        await this.resend();
+        break;
```
Comment (pinheadmz) [inline bin/hsw-cli:603]: Ok I didn't realize before but we have duplicate `resend` commands, with two functions in `hsw-cli` `resend()` and `resendWallet(id)`

The admin command resends from all wallets and is documented: https://hsd-dev.org/api-docs/#wallet-resend

in hsd: `this.post('/resend' ...`

The non-admin command requires a wallet id (or defaults to `primary`), and it is not documented.

in hsd: `this.post('/wallet/:id/resend ...`

So I think because of the admin shift/unshift thing the per-wallet resend was just never used (or not even accessible?) SO GOOD NEWS --> you get to make up a name for the command that just executes the resend function on one specific wallet :-)

File: bin/hsw-cli [hunk 4]
```diff
@@ -592,93 +592,76 @@ class CLI {
...
-        this.log('  $ block [height]: View wallet block.');
+        this.log('  $ unlock [passphrase] [timeout?]: Unlock wallet.');
         this.log('  $ view [tx-hex]: Parse and view transaction.');
-        this.log('  $ import [wif|hex]: Import private or public key.');
         this.log('  $ watch [address]: Import an address.');
-        this.log('  $ key [address]: Get wallet key by address.');
-        this.log('  $ dump [address]: Get wallet key WIF by address.');
-        this.log('  $ lock: Lock wallet.');
-        this.log('  $ unlock [passphrase] [timeout?]: Unlock wallet.');
+        this.log('  $ zap [age?]: Zap pending wallet TXs.');
```
Comment (pinheadmz) [inline bin/hsw-cli:645]: The `age` argument is actually required, I don't know why the `?` was ever there.

File: bin/hsw-cli [hunk 5]
```diff
@@ -592,93 +592,76 @@ class CLI {
...
+        this.log('');
+        this.log('Admin commands, they require api-token if node is run with wallet-wallet-auth flag.');
```
Comment (pinheadmz) [inline bin/hsw-cli:647]: Actually it requires `admin-token` and I think we can say here it is the `wallet-auth` flag, even though I know it may be `wallet-wallet-auth` -- perhaps link to the config guide? Like how you linked to API-docs below?

There is another layer which I don't know how to include here which is if wallet-auth is true, then each individual wallet requires its own wallet-token, for ANY command (the admin token overrides all individual wallet tokens). See https://hsd-dev.org/api-docs/#wallet-authentication

File: bin/hsw-cli [hunk 6]
```diff
@@ -592,93 +592,76 @@ class CLI {
...
+        this.log('Otherwise they can be called as the above ones, without any additional parameter.');
```
Comment (pinheadmz) [inline bin/hsw-cli:648]: I don't think this line is necessary as long as the first line is clear that if auth is set, token is required.

File: bin/hsw-cli [hunk 7]
```diff
@@ -592,93 +592,76 @@ class CLI {
...
-        this.log('  $ balance: Get wallet balance.');
-        this.log('  $ history: View TX history.');
-        this.log('  $ pending: View pending TXs.');
-        this.log('  $ coins: View wallet coins.');
-        this.log('  $ account list: List account names.');
+        this.log('  $ abandon [hash]: Abandon a transaction.');
         this.log('  $ account create [account-name]: Create account.');
         this.log('  $ account get [account-name]: Get account details.');
+        this.log('  $ account list: List account names.');
         this.log('  $ address: Derive new address.');
```
Comment (pinheadmz) [inline bin/hsw-cli:619]: `address` and `change` require an `account` parameter, wanna add that? (It is required deep in hsd/wallet.js ... probably should be checked in wallet/http.js but 🤷‍♂️ )

File: bin/hsw-cli [hunk 8]
```diff
@@ -592,93 +592,76 @@ class CLI {
...
+        this.log('  $ get: View wallet.');
+        this.log('  $ history: View TX history.');
+        this.log('  $ import [wif|hex]: Import private or public key.');
+        this.log('  $ key [address]: Get wallet key by address.');
+        this.log('  $ listen: Listen for events.');
+        this.log('  $ lock: Lock wallet.');
+        this.log('  $ master: View wallet master key.');
+        this.log('  $ mktx [address] [value]: Create transaction.');
+        this.log('  $ pending: View pending TXs.');
+        this.log('  $ rescan [height]: Rescan for transactions.');
```
Comment (pinheadmz) [inline bin/hsw-cli:635]: `master` and `rescan` require admin, I'm not sure why it wasn't already in that section...

**Review by pinheadmz — CHANGES_REQUESTED** [review]
Thanks @randomlogin this is looking good, I've just uncovered a few more mysteries around required parameters and admin stuff...

File: bin/hsw-cli (hunk 7 above)
Comment (randomlogin) [reply bin/hsw-cli:619]: I'm for using 'default' if it's not provided explicitly.

File: bin/hsw-cli (hunk 6 above)
Comment (randomlogin) [reply bin/hsw-cli:648]: Ok. I really hope it will be clear from the 1st line.

File: bin/hsw-cli (hunk 8 above)
Comment (randomlogin) [reply bin/hsw-cli:635]: Moved

File: bin/hsw-cli (hunk 4 above)
Comment (randomlogin) [reply bin/hsw-cli:645]: removed ?

File: bin/hsw-cli (hunk 5 above)
Comment (nodar-chkuaselidze) [reply bin/hsw-cli:647]: Actually, the `token` is used for both: admin authentication and wallet authentication. It's just if the `token === adminToken` then you are admin, otherwise you have to provide wallet specific `token`. So in both cases `token` is used.

There's no such thing as `api-token` or `admin-token` in parameters (admin-token only exists in wallet config file)

File: bin/hsw-cli [hunk 9]
```diff
@@ -592,93 +592,76 @@ class CLI {
...
         this.log('  $ resend: Resend pending transactions.');
-        this.log('  $ rescan [height]: Rescan for transactions.');
-        this.log('Other Options:');
+        this.log('  $ wallets: List all wallets.');
+        this.log('');
+        this.log('Other options:');
+        this.log('  --id [wallet id]: Wallet id.');
         this.log('  --passphrase [passphrase]: For signing/account-creation.');
         this.log('  --account [account-name]: Account name.');
+        this.log('  --api-token [token]: API or Admin token to run commands');
```
Comment (nodar-chkuaselidze) [inline bin/hsw-cli:659]: Same as above. This is just `--token [32 byte hex string]: Admin or Wallet token when walletAuth is true.`

**Review by nodar-chkuaselidze — CHANGES_REQUESTED** [review]

File: bin/hsw-cli (hunk 9 above)
Comment (nodar-chkuaselidze) [reply bin/hsw-cli:659]: Do we want to mention basic configs here?
  - Wallet http server configs:
    - `--http-host`, `--http-port`(or alternatively `--url`)
    - `--ssl`
    - `--api-key` - Basic auth for the wallet http server
  - Wallet node configs
    - `--network`?

File: bin/hsw-cli (hunk 3 above)
Comment (randomlogin) [reply bin/hsw-cli:603]: Okay, the per-wallet command seems to be not so useful, as resend all is innocent in my opinion (if tx was already tried to broadcast, then the user was indeed willing to do it, so there is no harm doing it again). So it the command can be a bit cryptic comparing to just `resend` for all wallets.

Following the `mkwallet` I suggest `resendwallet`.

Comment (randomlogin) [reply bin/hsw-cli:603]: By the way, providing unexsiting id will not cause any error. 
As there is no incorrect id handling for wallet actions (https://github.com/handshake-org/hsd/blob/master/lib/wallet/http.js#L801)

File: bin/hsw-cli (hunk 9 above)
Comment (randomlogin) [reply bin/hsw-cli:659]: Added `--api-key`, but not the rest.
I think we need to decide, if we want the hint to be tiny or to list all the commands.
I'm for listing all the parameters (and api commands), however it will require more things to add there.

File: bin/hsw-cli (hunk 5 above)
Comment (randomlogin) [reply bin/hsw-cli:647]: Added.
I also think it's authorization, not authentication (sorry).

File: bin/hsw-cli [hunk 10]
```diff
@@ -168,7 +168,6 @@ class CLI {
   }
 
   async createAddress() {
-    const account = this.config.str([0, 'account']);
```
Comment (pinheadmz) [inline bin/hsw-cli:171]: Why did you remove this?

```
--> hsw-cli address
ReferenceError: account is not defined
```

**Review by pinheadmz — CHANGES_REQUESTED** [review]
Looks good except for the missing `account` param. Also can you open an issue/PR for this?

> By the way, providing unexsiting id will not cause any error.
> As there is no incorrect id handling for wallet actions (https://github.com/handshake-org/hsd/blob/master/lib/wallet/http.js#L801)

File: bin/hsw-cli (hunk 10 above)
Comment (randomlogin) [reply bin/hsw-cli:171]: i've pushed the commit with my commented lines by accident, then amended it  in a hurry, and removed not only my lines, but also this.

**Review by pinheadmz — APPROVED** [review]
ACK 8cc8f1d01e372d1c9bdc5e037a00777303485566
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 8cc8f1d01e372d1c9bdc5e037a00777303485566
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmBYysQACgkQ5+KYS2KJ
yTr5Xg/9FWDdx+HViW7Jz2lZLx15ZsEcAjothA3FoJpFdQw7HFfH9wJXg0nXcF4J
wl/otwS/UTe6Zu1BUFGoCpn2syVdFroBHAeiSeKpKgl8PY4HmFCV+RUUrp2d63/X
OPCkII6Fqv9PDy3TuerJBx57QlP5mAgMhJ7XtP+bkeEtY0Bz8rMSl0VBaoTC/xcK
w936Zt2I82SMANThzmQ2V8Wzmxa/BvFIiPfUudAD4jxKCTQbdauoz12vsFGrFUzo
KIPx4u7j5i1fjmhu3TuRSTmehlKddVhB151UG7MZfLvboYs8/BC0QCeMHEGxWVy2
Tu4eiZB2L+z+tGg6ynfb4rnAKGVsImO0zNo6ZDjS2im/SFF7+XhS+11tCcVew8Yx
mSshzqA28XL72ewcmR3tB4AWK0hXdrYIomYJaqfGjd7xtD9MBSUVh0uyjJNCzv10
HDawQOWvcDga6rcCdMUFzngDp9j3dl7TnbSt3Zj+DNgXGUj661fbiOEKbNkc7z0d
PencnUgNU74SEzqfCLn9/HJyDeJ4VGvkmydBya/1u+8tfliC1xpgZVTkkei6fK15
rfoScJUPlXHGgjttw9sQ0ZNIrO76xtJmjW4+x3rezY3ezopXjrb6ORWH8+eUWU94
tvmMXTUQHL+yRtVkvkCf6EBv4X9mPLLjKZ7UKVUwQtb/+E+pp3Q=
=tHr3
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

File: bin/hsw-cli [hunk 11]
```diff
@@ -586,99 +586,87 @@ class CLI {
...
-        this.log('  $ account list: List account names.');
-        this.log('  $ account create [account-name]: Create account.');
-        this.log('  $ account get [account-name]: Get account details.');
-        this.log('  $ address: Derive new address.');
-        this.log('  $ change: Derive new change address.');
         this.log('  $ retoken: Create new api key.');
         this.log('  $ send [address] [value]: Send transaction.');
-        this.log('  $ mktx [address] [value]: Create transaction.');
+        this.log('  $ shared add [xpubkey]: Add key to wallet.');
+        this.log('  $ resendwallet [id]: Resend pending transactions for a single wallet.');
```
Comment (nodar-chkuaselidze) [inline bin/hsw-cli:640]: Even though, currently `.eslintfiles` is not setup which I should believe should look like this:
```
bin/hsd-cli
bin/hsw-cli
lib/
```

640, 649, 650 files are failing lint (p.s. I caught `double label` with linter as well.) (max len 80)
Maybe split with newlines ?

We could do this in another PR as well, but linter is useful as in the `double label` case.

File: bin/hsw-cli [hunk 12]
```diff
@@ -586,99 +586,87 @@ class CLI {
...
         this.log('  $ watch [address]: Import an address.');
-        this.log('  $ key [address]: Get wallet key by address.');
-        this.log('  $ dump [address]: Get wallet key WIF by address.');
-        this.log('  $ lock: Lock wallet.');
-        this.log('  $ unlock [passphrase] [timeout?]: Unlock wallet.');
-        this.log('  $ resend: Resend pending transactions.');
+        this.log('  $ zap [age]: Zap pending wallet TXs.');
+        this.log('');
+        this.log('If node is run with wallet-auth flag, wallet commands require authorization token.');
+        this.log('Admin commands require admin permissions for provided authorization token:');
```
Comment (nodar-chkuaselidze) [inline bin/hsw-cli:650]: I am not sure if this sentence means what it should?

You don't have per-token admin permissions in the wallet, you just have a single admin token that you need to use for any `admin` related functionality.

File: bin/hsw-cli [hunk 13]
```diff
@@ -586,99 +586,87 @@ class CLI {
...
+        this.log('  $ backup [path]: Backup the wallet db.');
+        this.log('  $ master: View wallet master key.');
+        this.log('  $ mkwallet [id]: Create wallet.');
```
Comment (nodar-chkuaselidze) [inline bin/hsw-cli:653]: `mkwalle` actually does not need admin token.

File: bin/hsw-cli [hunk 14]
```diff
@@ -586,99 +586,87 @@ class CLI {
...
-        this.log('Other Options:');
+        this.log('  $ resend: Resend pending transactions for all wallets.');
+        this.log('  $ rpc [command] [args]: Execute RPC command.');
+        this.log('  $ wallets: List all wallets.');
+        this.log('');
+        this.log('Other options:');
+        this.log('  --id [wallet id]: Wallet id.');
         this.log('  --passphrase [passphrase]: For signing/account-creation.');
         this.log('  --account [account-name]: Account name.');
+        this.log('  --token [token]: Wallet-specific authorization token.');
```
Comment (nodar-chkuaselidze) [inline bin/hsw-cli:663]: OR admin token.

**Review by nodar-chkuaselidze — CHANGES_REQUESTED** [review]
Maybe we should reorder `switch/case` as well? dunno just a thought.

Another thought?
Maybe we could move `address` and `change` under `account`?
- `account address`
- `account change`
I am still not sure about CLI conventions around these command formats.

These thoughts are just for the future probably.

File: bin/hsw-cli (hunk 12 above)
Comment (randomlogin) [reply bin/hsw-cli:650]: https://hsd-dev.org/api-docs/#wallet-authentication
There are wallet-specific tokens which you need to provide, if wallet-auth is true.

Comment (randomlogin) [reply bin/hsw-cli:650]: I just meant by this lines that if wallet-auth is true, then you must provide token no matter what, if it happens to be 'non-admin' token, then you will not be able to call 'admin' commands.

File: bin/hsw-cli [hunk 15]
```diff
@@ -586,99 +586,87 @@ class CLI {
...
       case 'rescan':
         await this.rescan();
         break;
+      case 'wallets':
+        await this.getWallets();
+        break;
+      case 'mkwallet':
+        await this.createWallet();
+        break;
+      case 'resend':
```
[the remark this answers is not in the document]
Comment (randomlogin) [reply bin/hsw-cli:604]: you're right

File: bin/hsw-cli (hunk 13 above)
Comment (randomlogin) [reply bin/hsw-cli:653]: Yes, you're right. I treat it as a bug as malicious user can create a lot of wallets. Created PR: https://github.com/handshake-org/hsd/pull/581

File: bin/hsw-cli (hunk 14 above)
Comment (randomlogin) [reply bin/hsw-cli:663]: Yep.

**Comment by randomlogin** [comment]
> Maybe we should reorder `switch/case` as well? dunno just a thought.
Done.

File: bin/hsw-cli [hunk 16]
```diff
@@ -528,157 +495,176 @@ class CLI {
           await this.createAccount();
           break;
         }
-        if (this.argv[0] === 'get')
+        if (this.argv[0] === 'get') {
```
Comment (nodar-chkuaselidze) [inline bin/hsw-cli:498]: I believe this is against hsd code styleguide.

File: bin/hsw-cli [hunk 17]
```diff
@@ -528,157 +495,176 @@ class CLI {
...
+          this.argv.shift();
+          await this.addSharedKey();
+          break;
+        }
+        if (this.argv[0] === 'remove') {
+          this.argv.shift();
+          await this.removeSharedKey();
+          break;
+        }
+        if (this.argv[0] === 'list') {
```
Comment (nodar-chkuaselidze) [inline bin/hsw-cli:586]: Same here.

File: lib/wallet.js [hunk 18]
```diff
@@ -197,6 +197,9 @@ class WalletClient extends Client {
    */
 
   createWallet(id, options) {
+    if (id === null) {
```
Comment (nodar-chkuaselidze) [inline lib/wallet.js:200]: Here u need to use `==` to catch `undefined` as well. Also there's no need for braces (look at hsd codebase)

File: lib/wallet.js [hunk 19]
```diff
@@ -1322,6 +1325,9 @@ class Wallet extends EventEmitter {
    */
 
   createAccount(name, options) {
+    if (name === null) {
```
Comment (nodar-chkuaselidze) [inline lib/wallet.js:1328]: same as above.

**Review by nodar-chkuaselidze — CHANGES_REQUESTED** [review]

File: lib/wallet.js (hunk 18 above)
Comment (pinheadmz) [reply lib/wallet.js:200]: Sorry Alex I think I led you astray on this one

File: bin/hsw-cli [hunk 20]
```diff
@@ -528,157 +495,174 @@ class CLI {
...
-        this.log('  $ account create [account-name]: Create account.');
-        this.log('  $ account get [account-name]: Get account details.');
-        this.log('  $ address: Derive new address.');
-        this.log('  $ change: Derive new change address.');
+        this.log('  $ resendwallet [id]: Resend pending transactions for a single wallet.');
         this.log('  $ retoken: Create new api key.');
         this.log('  $ send [address] [value]: Send transaction.');
-        this.log('  $ mktx [address] [value]: Create transaction.');
+        this.log('  $ shared add [xpubkey]: Add key to wallet.');
+        this.log('  $ shared remove [xpubkey]: Remove key from wallet.');
```
Comment (pinheadmz) [inline bin/hsw-cli:637]: `shared list` is missing. Also these require account names and since they affect accounts really (not wallets) I think this should be updated like so:

```suggestion
        this.log('  $ shared add [account-name] [xpubkey]: Add key to account.');
        this.log('  $ shared remove [account-name] [xpubkey]: Remove key from account.');
        this.log('  $ shared list [account-name]: List keys in account.');
```

**Comment by pinheadmz** [comment]
Thanks for sticking with this, @randomlogin it's a really important change that every user is going to see, so I'm glad we're taking the time. I think the only last important change is the `shared list`, everything else looks great.

@nodar-chkuaselidze I think the lint errors can be fixed in a follow-up PR, maybe also including eslint config files

**Review by nodar-chkuaselidze — APPROVED** [review]
